File: src/entities/Entity.js
/**
 * Entity Base Class
 * Base class for all game entities
 */
class Entity {
    constructor(x, y) {
        this.x = x;
        this.y = y;
        this.width = 0.8; // In world units
        this.height = 0.8;
        this.speed = 2; // tiles per second
        this.active = true;
    }
    
    /**
     * Update entity
     */
    update(deltaTime, world) {
        // Override in subclasses
    }
    
    /**
     * Render entity
     */
    render(renderer, camera, isometricRenderer) {
        // Override in subclasses
    }
    
    /**
     * Move entity
     */
    move(dx, dy, world) {
        const newX = this.x + dx;
        const newY = this.y + dy;
        
        // Check if new position is valid
        const tile = world.getTile(Math.floor(newX), Math.floor(newY));
        if (tile && (tile.isWalkable() || tile.isWater())) {
            this.x = newX;
            this.y = newY;
            return true;
        }
        
        return false;
    }
    
    /**
     * Get screen position
     */
    getScreenPosition(tileWidth, tileHeight) {
        return IsometricUtils.worldToScreen(this.x, this.y, tileWidth, tileHeight);
    }
}

File: cpp/src/entities/Entity.cpp
#include "entities/Entity.h"

Entity::Entity(float x, float y)
    : position(x, y)
    , velocity(0.0f, 0.0f)
    , active(true)
{
}

Entity::~Entity() {
}

void Entity::update(float deltaTime, World* world) {
    (void)world; // Unused in base class
    // Update position based on velocity
    position += velocity * deltaTime;
}

void Entity::render(Renderer* renderer, IsometricRenderer* isoRenderer, Camera* camera) {
    (void)renderer; // Unused in base class
    (void)isoRenderer; // Unused in base class
    (void)camera; // Unused in base class
    // Base entity has no visual representation
    // Override in derived classes
}

void Entity::setPosition(float x, float y) {
    position.x = x;
    position.y = y;
}

void Entity::setVelocity(float vx, float vy) {
    velocity.x = vx;
    velocity.y = vy;
}
